### PR TITLE
feat: use bitmasks within MUB (#1274)

### PR DESCRIPTION
**Pull request by tustvold**
Alters the MUB to store bitmasks instead of vecs of options.

This leads to a couple of things:

* Halves the amount of memory needed to store most values :tada:
* 10% faster ingest performance
* 15% faster conversion to arrow (and this will be even more pronounced with string dictionaries)

One potentially controversial change is using InfluxColumnType within the MUB. My view on this is that when we add support for data outside the influx data model, we will extend this enum to be able to store it, or create a new enum which is a superset of the current enumeration and change to using that. Ultimately I wanted a type I could use within the MUB that covered the full range of types, could be used in exhaustive matches and cheaply compared, and this seemed to fit the bill?

### CODE REVIEW

File: data_types/src/partition_metadata.rs [hunk 1]
```diff
@@ -222,8 +222,8 @@ impl Statistics {
 }
 
 /// Summary statistics for a column.
-#[derive(Debug, Deserialize, Serialize, Clone, PartialEq)]
-pub struct StatValues<T: PartialEq + PartialOrd + Debug + Display + Clone> {
+#[derive(Debug, Deserialize, Serialize, Clone, PartialEq, Default)]
```
Comment (tustvold) [inline data_types/src/partition_metadata.rs:225]: These changes make StatValues significantly easier to use. In particular:

* It can be constructed empty and subsequently appended to
* The string handling is no longer special cased

File: internal_types/src/entry.rs [hunk 2]
```diff
@@ -445,6 +445,37 @@ impl<'a> Column<'a> {
             .expect("name must be present in flatbuffers Column")
     }
 
+    pub fn inner(&self) -> &entry_fb::Column<'a> {
```
Comment (tustvold) [inline internal_types/src/entry.rs:448]: This change makes it so that the MUB can get access to the underlying null mask

File: internal_types/src/entry.rs [hunk 3]
```diff
@@ -445,6 +445,37 @@ impl<'a> Column<'a> {
...
+        &self.fb
+    }
+
+    pub fn influx_type(&self) -> InfluxColumnType {
```
Comment (tustvold) [inline internal_types/src/entry.rs:452]: This encodes the mapping from the Entry's types to the Influx data model types used by IOx's schema plumbing. As mentioned in the PR description, once we support types outside the InfluxDB data model we may want to expand either this enum or create a new parent enumeration, but this should be sufficient for now?

File: internal_types/src/schema.rs [hunk 4]
```diff
@@ -617,10 +617,10 @@ impl From<&InfluxColumnType> for &'static str {
     }
 }
 
-impl ToString for InfluxColumnType {
-    fn to_string(&self) -> String {
+impl std::fmt::Display for InfluxColumnType {
```
Comment (tustvold) [inline internal_types/src/schema.rs:620]: Necessary so that it can be used in Error types. ToString is automatically derived from the Display impl

File: mutable_buffer/src/bitset.rs [hunk 5]
```diff
@@ -0,0 +1,264 @@
+use arrow_deps::arrow::buffer::Buffer;
```
Comment (tustvold) [inline mutable_buffer/src/bitset.rs:1]: There are potentially more performant ways to implement a packed bitset, but I couldn't find any commonly used upstream crates, so I implemented one manually. Happy to split into separate PR if that is helpful

File: mutable_buffer/src/column.rs [hunk 6]
```diff
@@ -1,340 +1,262 @@
...
-                for v in vals {
-                    if let Some(val) = v {
-                        match stats.as_mut() {
-                            Some(s) => s.update(val),
-                            None => stats = Some(StatValues::new(val)),
-                        }
-                    }
-                    values.push(v);
-                }
+    pub fn new(row_count: usize, column_type: InfluxColumnType) -> Self {
```
Comment (tustvold) [inline mutable_buffer/src/column.rs:59]: Column's are now created as empty and then appended to, this avoids duplicating the write handling logic

File: mutable_buffer/src/column.rs [hunk 7]
```diff
@@ -1,340 +1,262 @@
...
+        self.validate_schema(entry)?;
+
+        let row_count = entry.row_count;
+        if row_count == 0 {
+            return Ok(());
+        }
+
+        let mask = construct_valid_mask(entry)?;
+
+        match &mut self.data {
```
Comment (tustvold) [inline mutable_buffer/src/column.rs:131]: Each match is fairly formulaic, it extracts the typed data array from the flatbuffer, extends the MUB's array with placeholder values, and then iterates through every set bit in the valid mask setting the value at the corresponding index

File: mutable_buffer/src/column.rs [hunk 8]
```diff
@@ -343,27 +265,150 @@ impl Column {
...
+                    .add_buffer(data.iter().cloned().collect())
+                    .null_bit_buffer(nulls)
+                    .build();
+                Arc::new(UInt64Array::from(data))
             }
-            Self::Tag(v, stats) => mem::size_of::<DID>() * v.len() + mem::size_of_val(&stats),
-            Self::String(v, stats) => {
-                let string_bytes_size = v
+            ColumnData::String(data, _) => {
+                // TODO: Store this closer to the arrow representation
```
Comment (tustvold) [inline mutable_buffer/src/column.rs:323]: I'm not sure we have many users adding strings as fields, so leaving this for now, but it would be a fairly straightforward change to reuse the storage backend used by the Dictionary.

File: mutable_buffer/src/column.rs [hunk 9]
```diff
@@ -343,27 +265,150 @@ impl Column {
...
+            ColumnData::Bool(data, _) => {
+                let data = ArrayDataBuilder::new(DataType::Boolean)
+                    .len(data.len())
+                    .add_buffer(data.to_arrow())
+                    .null_bit_buffer(nulls)
+                    .build();
+                Arc::new(BooleanArray::from(data))
             }
+            ColumnData::Tag(data, _) => {
+                // TODO: Store this closer to the arrow representation
```
Comment (tustvold) [inline mutable_buffer/src/column.rs:341]: This is the forthcoming StringDictionary work :)

File: mutable_buffer/src/table.rs [hunk 10]
```diff
@@ -324,27 +262,10 @@ impl Table {
...
-                    if column_name == TIME_COLUMN_NAME {
-                        schema_builder.timestamp()
-                    } else {
-                        schema_builder.field(column_name, ArrowDataType::Int64)
-                    }
-                }
-                Column::U64(_, _) => schema_builder.field(column_name, ArrowDataType::UInt64),
-                Column::Bool(_, _) => schema_builder.field(column_name, ArrowDataType::Boolean),
-            };
+            schema_builder = schema_builder.influx_column(col.column_name, column.influx_type());
```
Comment (tustvold) [inline mutable_buffer/src/table.rs:267]: Arrow conversion logic is largely pushed down into the Column itself, as it now has all the type information necessary to handle it

**Comment by alamb** [comment]
Checking this out....

File: internal_types/src/entry.rs (hunk 3 above)
Comment (alamb) [reply internal_types/src/entry.rs:452]: Maybe you could make the API more "future proof" by making this return an `Option` - similar to how `Schema` handles fields in the underlying `arrow::Schema` that don't have a corresponding type.

```suggestion
    pub fn influx_type(&self) -> Option<InfluxColumnType> {
```

That way people who use this API  will now they should be prepared to handle non `InfluxColumnTypes` in the future

File: mutable_buffer/src/bitset.rs (hunk 5 above)
Comment (alamb) [reply mutable_buffer/src/bitset.rs:1]: I notice you didn't use the arrow [`Bitmap`](https://github.com/apache/arrow-rs/blob/master/arrow/src/bitmap.rs)  ​or [`BooleanBufferBuilder`](https://github.com/apache/arrow-rs/blob/master/arrow/src/array/builder.rs#L287)

But you did use the builders in the test https://github.com/influxdata/influxdb_iox/pull/1289/files#diff-4a8f75a780ee4909b52e7705948083f5c140774ace179d3084b8a3bb2cc123f2R252

Is there some reason the arrow builder can't be used directly?

I think a custom bitmask thingie is fine for now, but wonder if this is something we could/should add to the arrow repo?

File: mutable_buffer/src/bitset.rs [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
+    /// Appends `count` unset bits
+    pub fn append_unset(&mut self, count: usize) {
+        self.len += count;
+        let new_buf_len = (self.len + 7) >> 3;
+        self.buffer.resize(new_buf_len, 0);
+    }
+
+    /// Appends `count` values from the iterator of packed bits
```
Comment (alamb) [inline mutable_buffer/src/bitset.rs:34]:
```suggestion
    /// Appends `count` boolean values from the iterator of packed bits
```

File: mutable_buffer/src/bitset.rs [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+        Buffer::from(&self.buffer)
+    }
+
+    /// Returns the number of values stored in the bitset
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Returns the number of bytes used by this bitset
+    pub fn byte_size(&self) -> usize {
```
Comment (alamb) [inline mutable_buffer/src/bitset.rs:104]:
```suggestion
    pub fn byte_len(&self) -> usize {
```

Maybe? For consistency?

File: mutable_buffer/src/bitset.rs [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
+    #[test]
+    fn test_compact_bools() {
+        let bools = &[
+            false, false, true, true, false, false, true, false, true, false,
+        ];
+        let collected = compact_bools(bools);
+        let indexes: Vec<_> = iter_set_bools(bools).collect();
+        assert_eq!(collected.as_slice(), &[0b01001100, 0b00000001]);
```
Comment (alamb) [inline mutable_buffer/src/bitset.rs:157]: I  checked this bit-by-bit  and it looks correct to me

File: mutable_buffer/src/bitset.rs [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+        assert_eq!(indexes.as_slice(), &[2, 3, 6, 8])
+    }
+
+    #[test]
+    fn test_bit_mask() {
+        let mut mask = BitSet::new();
+
+        mask.append_bits(8, &[0xFF]);
```
Comment (alamb) [inline mutable_buffer/src/bitset.rs:165]: stylistically, why `0xFF` and not `0b11111111` (to be consistent with the rest of the tests)?

File: mutable_buffer/src/column.rs [hunk 15]
```diff
@@ -1,340 +1,262 @@
...
-                    }
-                    values.push(v);
-                }
-
-                Self::F64(
-                    values,
-                    stats.expect("can't insert f64 column with no values"),
-                )
+            InfluxColumnType::Field(InfluxFieldType::UInteger) => {
+                let mut data = Vec::new();
```
Comment (alamb) [inline mutable_buffer/src/column.rs:70]: I wonder if using `vec![row_count; 0]` would be any better or worse than resizing a Vec? I ask for my own edification

File: mutable_buffer/src/column.rs (hunk 8 above)
Comment (alamb) [reply mutable_buffer/src/column.rs:323]: I agree with both of these sentiments

File: mutable_buffer/src/column.rs [hunk 16]
```diff
@@ -343,27 +265,150 @@ impl Column {
...
+                InvalidNullMask {
+                    expected_bytes: buf_len,
+                    actual_bytes: data.len()
+                }
+            );
+
+            Ok(data
+                .iter()
+                .map(|x| {
+                    // Currently the bit mask is backwards
```
Comment (alamb) [inline mutable_buffer/src/column.rs:377]: @pauldix  should we change the bitmasks in the flatbuffers structures to match Arrow? Namely arrow uses "validity" masks (1 where the row has a value, 0 where it is null) where as a more traditional "null mask" (which the flatbuffers structure has) uses 0 for a row with a value and 1 for a row with a null.

File: mutable_buffer/src/bitset.rs [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
+    /// Returns the number of bytes used by this bitset
+    pub fn byte_size(&self) -> usize {
+        self.buffer.len()
+    }
+}
+
+/// Returns an iterator over set bit positions in increasing order
+pub fn iter_set_bits(bytes: &[u8]) -> impl Iterator<Item = usize> + '_ {
```
Comment (alamb) [inline mutable_buffer/src/bitset.rs:110]: I was momentarily confused about this function as it is iterating over positions not bits (at least not what I think of as bits). I wonder if calling it something like

```suggestion
pub fn iter_set_positions(bytes: &[u8]) -> impl Iterator<Item = usize> + '_ {
```

Would make this clearer 🤔

File: mutable_buffer/src/column.rs [hunk 18]
```diff
@@ -343,27 +265,150 @@ impl Column {
...
+        data_size + self.valid.byte_size()
+    }
+
+    pub fn to_arrow(&self, dictionary: &Dictionary) -> Result<ArrayRef> {
+        let nulls = self.valid.to_arrow();
+        let data: ArrayRef = match &self.data {
+            ColumnData::F64(data, _) => {
+                let data = ArrayDataBuilder::new(DataType::Float64)
+                    .len(data.len())
+                    .add_buffer(data.iter().cloned().collect())
```
Comment (alamb) [inline mutable_buffer/src/column.rs:289]: the `clone()` is unfortunate perhaps, but this looks as good as the current state of affairs. So 👍

File: mutable_buffer/src/table.rs [hunk 19]
```diff
@@ -476,15 +347,15 @@ mod tests {
         ];
 
         write_lines_to_table(&mut table, &mut dictionary, lp_lines.clone());
-        assert_eq!(112, table.size());
+        assert_eq!(84, table.size());
```
Comment (alamb) [inline mutable_buffer/src/table.rs:350]: well the reported size is going in the correct direction too 👍

**Review by alamb — APPROVED** [review]
I think this is a really nice change @tustvold  -- nice work. 👍 

I think it would be worth explaining why this PR isn't using the Arrow bitmask / bitset.  Everything else is minor / code cleanliness suggestions.


> One potentially controversial change is using InfluxColumnType within the MUB.

I think the MUB relies on the Influx data model even before this PR  (implicitly) as it treats tags differently than String columns.

Using the `InfluxColumnType` to make it explicit and optimize that use case I think is very legit.

File: mutable_buffer/src/column.rs (hunk 18 above)
Comment (tustvold) [reply mutable_buffer/src/column.rs:289]: FWIW it is just "cloning" floats, but yes ideally it could just return an array pointing to the same backing memory without having to copy it out at all

File: mutable_buffer/src/column.rs (hunk 15 above)
Comment (tustvold) [reply mutable_buffer/src/column.rs:70]: I think they should be the same, will change as what you suggest is cleaner

File: mutable_buffer/src/bitset.rs (hunk 13 above)
Comment (tustvold) [reply mutable_buffer/src/bitset.rs:157]: You have my sympathy :sweat_smile:

File: internal_types/src/entry.rs (hunk 3 above)
Comment (tustvold) [reply internal_types/src/entry.rs:452]: At the moment this would just move the panic, as the MUB doesn't know how to handle a type outside the Influx data model. Unless you feel strongly I'd be inclined to leave it as is and revisit once that story is more fleshed out?

File: mutable_buffer/src/column.rs (hunk 15 above)
Comment (alamb) [reply mutable_buffer/src/column.rs:70]: I think it is fine as it is (no changes needed)

File: mutable_buffer/src/bitset.rs (hunk 5 above)
Comment (tustvold) [reply mutable_buffer/src/bitset.rs:1]: The limitation with the Builders as they stand at the moment is they're currently consuming - it currently isn't possible to pull data out of them and continue to perform writes. If this limitation were somehow addressed, then yes large portions of the MUB could theoretically move to using the builders but this doesn't appear to be the case currently? As for Bitmap, it appears to be immutable?

So the short answer is, yes the current builders can't be used as a drop-in for the MUB's logic.

That being said I think a form of mutable, snapshottable builder would be highly useful to have in arrow, and is effectively what the MUB is, but I'm not entirely sure what it would look like. Perhaps it would pre-allocate a contiguous block of memory up front and just return windows into that. For fixed length primitives this is relatively straightforward, but becomes more complex for strings, etc...

My view is it's probably worth iterating within IOx and with its specific constraints, and once we have something that's working well, maybe look to contribute it back up. I think keeping the MUB at least somewhat arrow agnostic will be advantageous for experimentation
